overlord/snapshotstate/snapshotstate_test.go

```diff
@@ -0,0 +1,662 @@
+		"snap":     "a-snap",
+		"filename": shotfile.Name(),
+		"users":    []interface{}{"a-user"},
+	})
+}
+
+func (snapshotSuite) TestCheckChecksIterError(c *check.C) {
+	defer snapshotstate.MockBackendIter(func(context.Context, func(*backend.Reader) error) error {
+		return errors.New("bzzt")
+	})()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	_, _, err := snapshotstate.Check(st, 42, nil, nil)
+	c.Assert(err, check.ErrorMatches, "bzzt")
+}
+
+func (snapshotSuite) TestCheckChecksForgetConflicts(c *check.C) {
+	shotfile, err := os.Create(filepath.Join(c.MkDir(), "yadda.zip"))
+	c.Assert(err, check.IsNil)
+	defer shotfile.Close()
+	fakeIter := func(_ context.Context, f func(*backend.Reader) error) error {
+		c.Assert(f(&backend.Reader{
+			// not wanted
+			Snapshot: client.Snapshot{SetID: 42, Snap: "a-snap"},
+			File:     shotfile,
+		}), check.IsNil)
+
+		return nil
+	}
+	defer snapshotstate.MockBackendIter(fakeIter)()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+	chg := st.NewChange("forget-snapshot", "...")
+	tsk := st.NewTask("some-task", "...")
+	tsk.SetStatus(state.DoingStatus)
+	tsk.Set("snapshot", map[string]int{"set-id": 42})
+	chg.AddTask(tsk)
+
+	_, _, err = snapshotstate.Check(st, 42, nil, nil)
+	c.Assert(err, check.ErrorMatches, `snapshot set #42 has a "forget-snapshot" change in progress`)
+}
+
+func (snapshotSuite) TestCheck(c *check.C) {
+	shotfile, err := os.Create(filepath.Join(c.MkDir(), "yadda.zip"))
+	c.Assert(err, check.IsNil)
+	defer shotfile.Close()
+	fakeIter := func(_ context.Context, f func(*backend.Reader) error) error {
+		c.Assert(f(&backend.Reader{
+			// not wanted
+			Snapshot: client.Snapshot{SetID: 42, Snap: "a-snap"},
+			File:     shotfile,
+		}), check.IsNil)
+
+		return nil
+	}
+	defer snapshotstate.MockBackendIter(fakeIter)()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	found, taskset, err := snapshotstate.Check(st, 42, []string{"a-snap", "b-snap"}, []string{"a-user"})
+	c.Assert(err, check.IsNil)
+	c.Check(found, check.DeepEquals, []string{"a-snap"})
+	tasks := taskset.Tasks()
+	c.Assert(tasks, check.HasLen, 1)
+	c.Check(tasks[0].Kind(), check.Equals, "check-snapshot")
+	c.Check(tasks[0].Summary(), check.Equals, `Check the data of snap "a-snap" in snapshot set #42`)
+	var snapshot map[string]interface{}
+	c.Check(tasks[0].Get("snapshot", &snapshot), check.IsNil)
+	c.Check(snapshot, check.DeepEquals, map[string]interface{}{
+		"set-id":   42.,
+		"snap":     "a-snap",
+		"filename": shotfile.Name(),
+		"users":    []interface{}{"a-user"},
+	})
+}
+
+func (snapshotSuite) TestForgetChecksIterError(c *check.C) {
+	defer snapshotstate.MockBackendIter(func(context.Context, func(*backend.Reader) error) error {
+		return errors.New("bzzt")
+	})()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	_, _, err := snapshotstate.Forget(st, 42, nil)
+	c.Assert(err, check.ErrorMatches, "bzzt")
+}
+
+func (snapshotSuite) TestForgetChecksCheckConflicts(c *check.C) {
```

Comment:
The various *Conflict checks tend to have a similar pattern, I wonder if some common helper could be extracted or even a table driven test be constructed.

Reply:
I resisted the urge of making things harder to follow :-) but maybe I drew the line too early.

Reply:
Yeah, its fine I was mostly wondering it may well be that consolidating it makes it actually (much) harder to read.